freenas/cli/plugins/openvpn.py

```diff
@@ -0,0 +1,473 @@
+
+        if kwargs['key_type'] == 'dh-parameters' and kwargs['key_length'] not in [1024, 2048]:
+            raise CommandException(_("wrong key_length value, see 'help generate_cryto' for more information"))
+
+        context.submit_task('service.openvpn.gen_key', kwargs['key_type'], kwargs.get('key_length'))
+
+@description("Provides corresponding OpenVPN client configuration")
+class OpenVPNClientConfigCommand(Command):
+    """
+    Usage: provide_client_config
+
+    Provides corresponding OpenVPN client configuration.
+    It needs to be copied to the OpenVPN client machine.
+    """
+    def run(self, context, args, kwargs, opargs):
+
+        vpn_client_confg = context.call_sync('service.openvpn.client_config.provide_config')
+        return Sequence(vpn_client_confg)
+
+@description("Provides OpenVPN static key")
+class OpenVPNStaticKeyCommand(Command):
+    """
+    Usage: provide_static_key
+
+    Provides static OpenVPN key that needs to be copied to the client machine
+    """
+    def run(self, context, args, kwargs, opargs):
+
+        vpn_client_confg = context.call_sync('service.openvpn.get_config')
+        return Sequence(vpn_client_confg['tls_auth'])
+
+
+@description("Manages OpenVPN service")
+class OpenVPNManageNamespace(EntitySubscriberBasedLoadMixin, EntityNamespace):
+    """
+    This namespace provides commands for managing the OpenVPN service.
+    """
+    def __init__(self, name, context):
+        super(OpenVPNManageNamespace, self).__init__(name, context)
+
+        self.entity_subscriber_name = 'service'
+        self.extra_query_params = [('name', '=', 'openvpn')]
+        self.primary_key = self.get_mapping('name')
+
+        self.allow_create = False
+        self.allow_edit = False
+
+        self.add_property(
+            descr='Service name',
+            name='name',
+            get='name',
+            usage=_("""\
+            Name of the service. Read-only value assigned by
+            the operating system."""),
+            set=None,
+            list=True
+        )
+
+        self.add_property(
+            descr='State',
+            name='state',
+            get='state',
+            usage=_("""\
+            Indicates whether the service is RUNNING or STOPPED.
+            Read-only value assigned by the operating system."""),
+            set=None,
+            list=True
+        )
+
+        self.add_property(
+            descr='Process ID',
+            name='pid',
+            get='pid',
+            usage=_("""\
+            Process ID of the RUNNING service. Read-only value assigned
+            by the operating system."""),
+            set=None,
+            list=True
+        )
+
+        self.extra_commands = {
+            'start': OpenVPNManageCommand('start'),
+            'stop': OpenVPNManageCommand('stop'),
+            'reload': OpenVPNManageCommand('reload'),
+            'restart': OpenVPNManageCommand('restart')
+        }
```

Comment:
Yes - I' didn't wanted to change the structure of the services plugin.
It is designed in polymorfic way and I didn't had an idea how to change it in elegant way.